### PR TITLE
update group calibration script

### PR DESCRIPTION
**Pull request by Kvieta1990**
**Description of work.**

1. More input control parameters added to the group calibration routine, including peak function type for estimating offset after cross correlation and an option to turn on or off the smoothing of data for cross correlation purpose. The workflow of group calibration script is also polished to make it smoother. Accordingly, unit tests have been updated.

2. Fix the issue with the calibration diagnostics script when dealing with instruments of which the detector ID does not start from 0.

**To test:**

Run the group_calib_test.py script with all inputs provided in the attached package, [group_calib.zip](https://github.com/mantidproject/mantid/files/6993217/group_calib.zip)

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/Calibration/tofpd/group_calibration.py [hunk 1]
```diff
@@ -291,11 +308,14 @@ def pdcalibration_groups(data_ws,
...
-        difc_calib = mtd[f'{output_basename}_cc_pd_diffcal_tmp'].row(i)['difc']
+        if i in to_skip:
+            difc_calib = mtd[f'{output_basename}_pd_diffcal_skip'].row(i)['difc']
+        else:
+            difc_calib = mtd[f'{output_basename}_cc_pd_diffcal_tmp'].row(i)['difc']
         if mtd[f'{output_basename}_pd_diffcal_mask'].readY(i)[0] == 0.0:
             diff_difc = abs(difc_bak - difc_calib) / difc_calib * 100.0
         else:
-            diff_difc = 0.0
+            diff_difc = float("inf")
```
Comment (peterfpeterson) [inline scripts/Calibration/tofpd/group_calibration.py:318]:
```suggestion
            diff_difc = numpy.inf
```

File: scripts/test/Calibration/tofpd/test_group_calibration.py [hunk 2]
```diff
@@ -62,32 +65,49 @@ def test_from_eng(self):
...
 
-        assert_equal(mask.extractY(), [[0],
-                                       [0],
-                                       [0],
-                                       [1],
-                                       [0],
-                                       [0],
-                                       [0],
-                                       [1]])
+        assert_equal(mtd[f'{output_workspace_basename}_pd_diffcal_mask'].extractY(), [[0],
```
Comment (peterfpeterson) [inline scripts/test/Calibration/tofpd/test_group_calibration.py:103]: The assertion error, when it happens, will be clearer if you use `numpy.testing.assert_equal`

File: scripts/test/Calibration/tofpd/test_group_calibration.py [hunk 3]
```diff
@@ -150,14 +171,14 @@ def test_from_prev_cal(self):
...
 
-        assert_equal(mask.extractY(), [[0],
-                                       [0],
-                                       [0],
-                                       [1],
-                                       [0],
-                                       [0],
-                                       [0],
-                                       [1]])
+        assert_equal(mtd['group_calibration_pd_diffcal_mask'].extractY(), [[0],
```
Comment (peterfpeterson) [inline scripts/test/Calibration/tofpd/test_group_calibration.py:174]: The assertion error, when it happens, will be clearer if you use `numpy.testing.assert_equal`

File: scripts/test/Calibration/tofpd/test_group_calibration.py [hunk 4]
```diff
@@ -206,14 +227,14 @@ def test_di_group_calibration(self):
...
```
Comment (peterfpeterson) [inline scripts/test/Calibration/tofpd/test_group_calibration.py:230]: The assertion error, when it happens, will be clearer if you use `numpy.testing.assert_equal`

File: scripts/Calibration/tofpd/group_calibration.py [hunk 5]
```diff
@@ -177,16 +190,16 @@ def cc_calibrate_groups(data_ws,
     DeleteWorkspace('_accum_cc')
     DeleteWorkspace('_tmp_group_cc')
     DeleteWorkspace('_tmp_group_cc_raw')
-    if cycling:
+    if cycling and cycled:
```
Comment (peterfpeterson) [inline scripts/Calibration/tofpd/group_calibration.py:193]: Could the `cycled` variable be removed and replace this check with
```python
if cycling and '_tmp_group_cc_diffcal' in mtd:
```
to capture the intent?

**Review by peterfpeterson — COMMENTED** [review]
The while loop in the current group_calibration.py#128 can be modified using `from itertools import count`
```python
for num_cycle in count(start=1, step=1):
```
then the related code for updating the variable can be removed.

File: scripts/test/Calibration/tofpd/test_group_calibration.py (hunk 2 above)
Comment (Kvieta1990) [reply scripts/test/Calibration/tofpd/test_group_calibration.py:103]: I think we are just using `numpy.testing.assert_equal` here -- if you look at the `import` statement at the top of the script, we were doing `from numpy.testing import assert_allclose, assert_equal` there.

File: scripts/test/Calibration/tofpd/test_group_calibration.py (hunk 3 above)
Comment (Kvieta1990) [reply scripts/test/Calibration/tofpd/test_group_calibration.py:174]: I think we are just using numpy.testing.assert_equal here -- if you look at the import statement at the top of the script, we were doing from numpy.testing import assert_allclose, assert_equal there.

File: scripts/test/Calibration/tofpd/test_group_calibration.py (hunk 4 above)
Comment (Kvieta1990) [reply scripts/test/Calibration/tofpd/test_group_calibration.py:230]: I think we are just using numpy.testing.assert_equal here -- if you look at the import statement at the top of the script, we were doing from numpy.testing import assert_allclose, assert_equal there.

**Review by peterfpeterson — DISMISSED** [review]

**Review by peterfpeterson — APPROVED** [review]

**Pull request merged**